extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js

```diff
@@ -13,3 +13,121 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+/**
+ * @fileoverview Service for recaptcha components
+ * interacting with the 3p recaptcha bootstrap iframe
+ */
+
+import {Deferred} from '../../../src/utils/promise';
+import {dev} from '../../../src/log';
+import {getIframe} from '../../../src/3p-frame';
+import {getService, registerServiceBuilder} from '../../../src/service';
+import {listenFor} from '../../../src/iframe-helper';
+import {removeElement} from '../../../src/dom';
+
+export class AmpRecaptchaService {
+
+  /**
+   * @param {!Window} window
+   */
+  constructor(window) {
+
+    /** @private {!Window} */
+    this.win_ = window;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {number} */
+    this.registeredElements_ = 0;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /** @private {!Deferred} */
+    this.willBeReady_ = new Deferred();
+  }
+
+  /**
+   * Function to register as a dependant of the AmpRecaptcha serivce.
+   * Used to create/destroy recaptcha boostrap iframe.
+   * @param {!AMP.BaseElement} elementImpl
+   * @return {Promise}
+   */
+  register(elementImpl) {
+    this.registeredElements_++;
+    if (!this.iframe_) {
+      this.initialize_(elementImpl);
+    }
+    return elementImpl.loadPromise(this.iframe_);
+  }
+
+  /**
+   * Function to unregister as a dependant of the AmpRecaptcha serivce.
+   * Used to create/destroy recaptcha boostrap iframe.
+   */
+  unregister() {
+    this.registeredElements_--;
+    if (this.registeredElements_ <= 0) {
+      this.dispose_();
+    }
+  }
+
+  /**
+   * Function to create our recaptcha boostrap iframe.
+   * @param {!AMP.BaseElement} elementImpl
+   * @private
+   */
+  initialize_(elementImpl) {
+
+    /* the third parameter 'recaptcha' ties it to the 3p/recaptcha.js */
+    this.iframe_ = getIframe(this.win_, elementImpl.element, 'recaptcha');
+
+    const listenIframe = (evName, cb) => listenFor(
+        dev().assertElement(this.iframe_),
+        evName,
+        cb,
+        true);
+
+    const disposers = [
+      listenIframe('ready', this.willBeReady_.resolve),
+    ];
+    this.unlistenMessage_ = () => disposers.forEach(d => d());
+
+    elementImpl.element.appendChild(this.iframe_);
```

Comment:
Curious have we considered about appending the iframe to the document? Would it be possible that the parent element get removed from DOM while we still want to keep the recaptcha iframe for other components.

Reply:
Let's talk offline about this 😄